# RNA-sequencing snakemake pipeline

Here we present a mRNA-sequencing `snakemake` pipeline that can be used for data processing, namely:
- alignment using STAR
- post-read processing using SAMtools to remove bad alignments and mitochondrial reads
- UMI-demultiplexing using UMI-tools
- read counting/mapping using Subread::featureCount

For this pipeline to work, it is assumed that you have [snakemake installed](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html). Furthermore, this pipeline makes use of conda environments. Finally, this pipeline has been optimized for data generated by the AmsterdamUMC Core Facility Genomics. Data obtained from other sequencing facilities might not work out-of-the-box and might require some tooling.

Use this pipeline at your own risk!

## Quickstart

You should first setup the conda environment as defined in the `environment.yml` file, which will install `snakemake`, `Python`, as well as the Python packages `pandas` and `openpyxl`.

```
conda env create -f environment.yml
```

To **test** this pipeline, use.

```
snakemake --cores [cores] --use-conda -np
```

This pipeline was run using the following command

```
snakemake --cores 20 -p --use-conda --resources mem_mb=189000 --rerun-triggers mtime
```